enhancements/network/external-lb-vips.md

```diff
@@ -0,0 +1,213 @@
+   will not set the fields for apiVIP and ingressVIP. We should fail if they set these mutually exclusive parameters.
+4. In the installer, if the platform does support external loadbalancer, some validation will be run
+   to check that the VIPs can already be resolved (the validation already exists for vsphere, we need to
+    make it re-usable). A new field in `PlatformStatus` will be set.
+5. In machine-config-operator, if LB type is external, the internal loadbalancer will not be
+   deployed.
+6. When using an external loadbalancer, the VIPs fields won't be set in `PlatformStatus` and the
+   OpenShift components won't use it when they're not present and LB type is external.
+7. After the deployment, the IP address of a service (e.g. API) can change, as long as the external
+   loadbalancer is updated and the DNS record has been updated and the results are propagated as well.
+
+### API Extensions
+
+A new field will be added to the `PlatformStatus` of the infrastructure object to tell
+the kind of the loadbalancer.
+
+```golang
+// LoadBalancerType can be either external or internal 
+type LoadBalancerType string
+const (
+	// LoadBalancerTypeExternal
+	LoadBalancerTypeExternal LoadBalancerType = "external"
+	// LoadBalancerTypeInternal
+	LoadBalancerTypeInternal LoadBalancerType = "internal"
+)
+// VSpherePlatformStatus holds the current status of the vSphere infrastructure provider.
+type VSpherePlatformStatus struct {
+  // loadBalancerType defines if a internal or external load balancer
+  // is to be used for the openshift api and ingress endpoints.
+  // When set to Internal the static pods defined in machine config operator
+  // will be used.
+  // When set to External the static pods will not be deployed. It will be 
+  // expected that the LB and DNS A records be pre-configured prior to installation.
+  // +kubebuilder:default=internal
+  // +kubebuilder:validation:Enum=internal;external
+  LoadBalancer LoadBalancerType `json:"loadBalancerType,omitempty"`
+}
+```
+
+This field will be set in the `install-config.yaml`:
+```yaml
+platform:
+  vsphere:
+    loadBalancerType: external
+```
+
+### Implementation Details/Notes/Constraints
+
+* CoreDNS will not be deployed if external loadbalancer is being used.
+  The customer is responsible of populating the DNS records for api, api-int and ingress.
+  We need to check who is responsible for machine records.
+
+* We need to make it ready for any platform who wants to support it, and allow
+  platforms to turn this on in the installer project. This will be achieved while validating the
+  installconfig manifest.
+
+* `PlatformStatus` won't have the VIP fields if LB type is external.
+
+
+### Risks and Mitigations
+
+Minimal risk, some platforms (vsphere, openstack) have been testing it already and no issue
+has been seen so far with this scenario.
+Validations will have to be implemented to make sure that the customer has prepare their
+environment before the IPI deployment.
```

Comment:
Having an external LB breaks [`nodeip-finder`](https://github.com/openshift/baremetal-runtimecfg/blob/master/scripts/nodeip-finder). This results in kubelet setting an empty node-ip. This in turn means that the CCM (or kubelet in the case of legacy cloud provider) will does not filter node addresses, and will expose a kubelet endpoint on all local interfaces. This is undesirable and may be a security issue in certain cases, for example when nodes are attached to a public-facing dataplane network.

Reply:
@mdbooth  I'll add that in the risks. Do you have a link on the work that is happening to mitigate this issue? I know you looked at it with @MaysaMacedo . Thanks.

Reply:
@EmilienM Is a JIRA card what you're looking for?

Reply:
@MaysaMacedo anything we can link it here for reference.

Reply:
nodeip-configuration should work without a VIP, it just falls back to the UPI/remote worker path where it selects based on default route. We could add support for NODEIP_HINT in this use case too (in fact we probably should).

Reply:
@cybertron I think the script that it's broken is `[node-ip set](https://github.com/openshift/machine-config-operator/blob/master/templates/common/on-prem/units/nodeip-configuration.service.yaml#L28-L29)`. In any case, your suggestion is to enable [this script](https://github.com/openshift/machine-config-operator/blob/master/templates/common/_base/units/nodeip-configuration.service.yaml#L2) for on-premise platforms?

(A bit out of context of this PR) Given that it accepts only one IP, does this also work well for dual-stack environments?

Reply:
I'm saying that service still works fine even if it can't find the VIP or if one isn't provided. That's what happens on remote workers and UPI deployments. I suppose it's possible there's a syntax error if the VIP isn't there since this specific one was written assuming it always would be, but that should be easily fixed.

It also accepts multiple VIPs as of https://github.com/openshift/machine-config-operator/pull/3269 so dual stack is fully supported.

Reply:
I'll leave the comment open for now but I added it as a risk.